web3/_utils/contract_error_handling.py

```diff
@@ -49,67 +49,47 @@
+    """
+    prefix = f"Reverted {SOLIDITY_ERROR_FUNC_SELECTOR}"
+    data_offset = ("00" * 31) + "20"  # 0x0000...0020 (32 bytes)
+    revert_pattern = prefix + data_offset
+    error = data
 
-    data = response["error"].get("data", "")
-    message = response["error"].get("message", "")
+    if data.startswith(revert_pattern):
+        # if common revert pattern
+        string_length = int(data[len(revert_pattern) : len(revert_pattern) + 64], 16)
+        error = data[
+            len(revert_pattern) + 64 : len(revert_pattern) + 64 + string_length * 2
+        ]
+    elif data.startswith("Reverted 0x"):
+        # Special case for this form: 'Reverted 0x...'
+        error = data.split(" ")[1][2:]
 
-    message_present = message is not None and message != ""
+    return error
 
-    if data is None:
-        if message_present:
-            raise ContractLogicError(message)
-        elif not message_present:
-            raise ContractLogicError("execution reverted")
-        else:
-            raise Exception("Unreachable")
-
-    # Ganache case:
-    if isinstance(data, dict) and message_present:
-        raise ContractLogicError(f"execution reverted: {message}", data=data)
 
-    # Parity/OpenEthereum case:
+def _raise_error_from_decoded_revert_data(data: str) -> None:
+    """
+    Decode response error data and raise appropriate exception.
+    """
     if data.startswith("Reverted "):
-        # "Reverted", function selector and offset are always the same for revert errors
-        prefix = f"Reverted {SOLIDITY_ERROR_FUNC_SELECTOR}"
-        data_offset = ("00" * 31) + "20"  # 0x0000...0020 (32 bytes)
-        revert_pattern = prefix + data_offset
-
-        if data.startswith(revert_pattern):
-            # if common revert pattern
-            string_length = int(
-                data[len(revert_pattern) : len(revert_pattern) + 64], 16
-            )
-            reason_as_hex = data[
-                len(revert_pattern) + 64 : len(revert_pattern) + 64 + string_length * 2
-            ]
-        elif data.startswith("Reverted 0x"):
-            # Special case for this form: 'Reverted 0x...'
-            reason_as_hex = data.split(" ")[1][2:]
-        else:
-            raise ContractLogicError("execution reverted", data=data)
-
         try:
-            reason_string = bytes.fromhex(reason_as_hex).decode("utf8")
-            raise ContractLogicError(f"execution reverted: {reason_string}", data=data)
+            reason_string = bytes.fromhex(_parse_openethereum_hex_error(data)).decode(
+                "utf8"
+            )
         except UnicodeDecodeError:
             warnings.warn("Could not decode revert reason as UTF-8", RuntimeWarning)
             raise ContractLogicError("execution reverted", data=data)
 
-    # --- EIP-3668 | CCIP Read --- #
+        raise ContractLogicError(f"execution reverted: {reason_string}", data=data)
+
     if data[:10] == OFFCHAIN_LOOKUP_FUNC_SELECTOR:
```

Comment:
If they are mutually exclusive, I'd say use an `if-elif-else` instead of 3 `if`'s here. Is there a test that catches if none of the conditions are met?